##**Task 1**

In [17]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, MinMaxScaler

In [18]:
#  Load the dataset
df = pd.read_csv('/content/adult_with_headers (1).csv')

In [19]:
df

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32557,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32558,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K
32559,22,Private,201490,HS-grad,9,Never-married,Adm-clerical,Own-child,White,Male,0,0,20,United-States,<=50K


In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [21]:
df.isna().sum()

,0
age,0
workclass,0
fnlwgt,0
education,0
education_num,0
marital_status,0
occupation,0
relationship,0
race,0
sex,0


In [22]:
df.isnull().sum()

,0
age,0
workclass,0
fnlwgt,0
education,0
education_num,0
marital_status,0
occupation,0
relationship,0
race,0
sex,0


In [23]:
df.describe()

,age,fnlwgt,education_num,capital_gain,capital_loss,hours_per_week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [24]:
df.shape

(32561, 15)

In [25]:
# Handle missing values
# In the Adult dataset, missing values are often represented as ' ?'
# We replace these with NaN to make them easier to handle
df.replace(' ?', np.nan, inplace=True)
df.replace('?', np.nan, inplace=True)

In [26]:
# Identify categorical columns with missing values
categorical_cols_with_nan = ['workclass', 'occupation', 'native_country']

# Apply Mode Imputation: fill missing categorical values with the most frequent value
for col in categorical_cols_with_nan:
    df[col] = df[col].fillna(df[col].mode()[0])

In [27]:
# Define numerical features for scaling
numerical_features = ['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']

In [28]:
# Apply Scaling Techniques

# Standard Scaling (Z-score Normalization)
# Centers data around 0 with a standard deviation of 1
scaler_standard = StandardScaler()
df_standard = df.copy()
df_standard[numerical_features] = scaler_standard.fit_transform(df[numerical_features])

#  Min-Max Scaling (Normalization)
# Scales data to a fixed range [0, 1]
scaler_minmax = MinMaxScaler()
df_minmax = df.copy()
df_minmax[numerical_features] = scaler_minmax.fit_transform(df[numerical_features])

#  Save the results
df_standard.to_csv('adult_standard_scaled.csv', index=False)
df_minmax.to_csv('adult_minmax_scaled.csv', index=False)

# Optional: Verify the scaling results
print("--- Standard Scaled Numerical Stats ---")
print(df_standard[numerical_features].describe().loc[['mean', 'std', 'min', 'max']])

print("\n--- Min-Max Scaled Numerical Stats ---")
print(df_minmax[numerical_features].describe().loc[['mean', 'std', 'min', 'max']])

--- Standard Scaled Numerical Stats ---
               age        fnlwgt  education_num  capital_gain  capital_loss  \
mean -2.705915e-17 -1.001625e-16   1.471887e-16  1.309314e-17  1.016900e-16   
std   1.000015e+00  1.000015e+00   1.000015e+00  1.000015e+00  1.000015e+00   
min  -1.582206e+00 -1.681631e+00  -3.529656e+00 -1.459205e-01 -2.166595e-01   
max   3.769612e+00  1.226856e+01   2.300838e+00  1.339458e+01  1.059351e+01   

      hours_per_week  
mean   -1.549355e-17  
std     1.000015e+00  
min    -3.194030e+00  
max     4.742967e+00  

--- Min-Max Scaled Numerical Stats ---
           age    fnlwgt  education_num  capital_gain  capital_loss  \
mean  0.295639  0.120545       0.605379      0.010777      0.020042   
std   0.186855  0.071685       0.171515      0.073854      0.092507   
min   0.000000  0.000000       0.000000      0.000000      0.000000   
max   1.000000  1.000000       1.000000      1.000000      1.000000   

      hours_per_week  
mean        0.402423  
std    

## A. Standard Scaling (Z-score Normalization)
###How it works: It transforms data to have a mean of $0$ and a standard deviation of $1$.

###*When to prefer it:*

###Gaussian Distribution: When your features follow a Normal (Gaussian) distribution.

###Algorithm Assumptions: It is preferred for algorithms that assume the data is centered around zero or follow a normal distribution, such as Linear Regression, Logistic Regression, and Linear Discriminant Analysis (LDA).

###Distance/Variance Based: Essential for Principal Component Analysis (PCA) and Support Vector Machines (SVM) with RBF kernels.

###Outliers: It is generally more robust to outliers than Min-Max scaling, as it doesn't compress the majority of data into a tiny range due to a single extreme value.

##B. Min-Max Scaling (Normalization)

###How it works: It scales the data to a fixed range, usually $[0, 1]$, by subtracting the minimum and dividing by the range.

###*When to prefer it:*

###No Distribution Assumption: When you do not know the distribution of your data or if it is definitely not Gaussian.

###Neural Networks: Often used in Deep Learning as it helps gradients converge faster when inputs are in a small, bounded range.

###Magnitude Sensitivity: Useful for algorithms that are sensitive to the absolute scale of features, such as k-Nearest Neighbors (k-NN) and Recommender Systems.

###Image Processing: Standard practice for pixel intensities (scaling 0–255 to 0–1).Constraint: Should be avoided if there are significant outliers, as they will "squash" the normal data points into a very small interval (e.g., all values become 0.001 because one outlier is at 1.0).

##**Task 2**

In [29]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

# Load the dataset
df = pd.read_csv('/content/adult_with_headers (1).csv')

# Basic Data Exploration
print("--- Data Info ---")
print(df.info())

print("\n--- Summary Statistics ---")
print(df.describe())

print("\n--- Missing Values Check ---")
# Replacing ' ?' with NaN for accurate counting
df.replace(' ?', np.nan, inplace=True)
df.replace('?', np.nan, inplace=True)
print(df.isnull().sum())

# Handle Missing Values
# Impute categorical columns with the mode (most frequent value)
categorical_cols = df.select_dtypes(include=['object']).columns
for col in categorical_cols:
    df[col] = df[col].fillna(df[col].mode()[0])

--- Data Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB
None

--- Summary Statistics ---
                age        fnlwgt  educati

In [30]:
# Encoding Techniques based on Category Count
unique_counts = {col: df[col].nunique() for col in categorical_cols}

# Lists to store columns for each encoding type
ohe_cols = [col for col in categorical_cols if unique_counts[col] < 5]
label_cols = [col for col in categorical_cols if unique_counts[col] >= 5]

print("\nOne-Hot Encoding applied to:", ohe_cols)
print("Label Encoding applied to:", label_cols)



One-Hot Encoding applied to: ['sex', 'income']
Label Encoding applied to: ['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'race', 'native_country']


In [31]:
# Apply One-Hot Encoding (OHE)
# drop_first=True is often used to avoid the dummy variable trap
df_encoded = pd.get_dummies(df, columns=ohe_cols, drop_first=True)

# Apply Label Encoding
le = LabelEncoder()
for col in label_cols:
    df_encoded[col] = le.fit_transform(df_encoded[col])

# Review final processed data
print("\n--- Final Data Types ---")
print(df_encoded.dtypes)
print("\n--- Processed Data Head ---")
print(df_encoded.head())

# Save the encoded dataframe
df_encoded.to_csv('adult_encoded.csv', index=False)


--- Final Data Types ---
age               int64
workclass         int64
fnlwgt            int64
education         int64
education_num     int64
marital_status    int64
occupation        int64
relationship      int64
race              int64
capital_gain      int64
capital_loss      int64
hours_per_week    int64
native_country    int64
sex_ Male          bool
income_ >50K       bool
dtype: object

--- Processed Data Head ---
   age  workclass  fnlwgt  education  education_num  marital_status  \
0   39          6   77516          9             13               4   
1   50          5   83311          9             13               2   
2   38          3  215646         11              9               0   
3   53          3  234721          1              7               2   
4   28          3  338409          9             13               2   

   occupation  relationship  race  capital_gain  capital_loss  hours_per_week  \
0           0             1     4          2174             0  

#Discussion: Pros and Cons of Encoding Techniques

#**One-Hot Encoding**

##Pros:

No Ordinal Bias: It treats all categories as equal, which is ideal for nominal data (where there is no inherent ranking).

Distance Friendly: Works well for algorithms like k-Nearest Neighbors or SVMs that rely on calculating distances between points.

##Cons:

Dimensionality Issues: Can create a massive number of features if the category count is high, leading to increased memory usage and training time.

Multicollinearity: Can introduce redundancy (often solved by using drop_first=True).

#**Label Encoding**

##Pros:

Memory Efficient: Keeps the dataset compact by using a single column of integers.

Tree-Model Compatible: Decision trees and Random Forests can naturally handle integer labels by splitting the data at specific values.

##Cons:

Artificial Ranking: Linear models may misinterpret the integers as a ranked order (e.g., thinking a category assigned '4' is "greater than" a category assigned '1').

Not Ideal for Distance-Based Models: Can lead to poor results in models like k-Means clustering or Linear Regression because the numerical "distance" between labels is meaningless.
#

In [32]:
import pandas as pd
import numpy as np

# 1. Load and handle basic missing values first
df = pd.read_csv('/content/adult_with_headers (1).csv')
df.replace(' ?', np.nan, inplace=True)
df.replace('?', np.nan, inplace=True)
categorical_cols = df.select_dtypes(include=['object']).columns
for col in categorical_cols:
    df[col] = df[col].fillna(df[col].mode()[0])

# 2. Create New Features
# Feature 1: Net Capital Movement
df['net_capital'] = df['capital_gain'] - df['capital_loss']

# Feature 2: Work-Education Interaction
df['work_edu_index'] = df['hours_per_week'] * df['education_num']

print("--- New Features Created ---")
print(df[['net_capital', 'work_edu_index']].head())

# 3. Transformation for Skewed Features
# Checking skewness before transformation
original_skew = df['capital_gain'].skew()
print(f"\nSkewness of 'capital_gain' before transformation: {original_skew:.2f}")

# Applying Log Transformation (using log1p to handle 0 values safely)
df['capital_gain_log'] = np.log1p(df['capital_gain'])

# Checking skewness after transformation
new_skew = df['capital_gain_log'].skew()
print(f"Skewness of 'capital_gain' after Log transformation: {new_skew:.2f}")

# Save the updated dataset
df.to_csv('adult_feature_engineered.csv', index=False)

--- New Features Created ---
   net_capital  work_edu_index
0         2174             520
1            0             169
2            0             360
3            0             280
4            0             520

Skewness of 'capital_gain' before transformation: 11.95
Skewness of 'capital_gain' after Log transformation: 3.10


#Feature Engineering & Transformations
##New Features
##*Creatednet_capital:* Calculated as capital_gain - capital_loss.
* Rationale: Instead of treating gains and losses as independent events, the net value represents the actual financial growth or deficit for an individual. This provides a more holistic view of a person's financial status in a single feature.

##*work_edu_index:* Calculated as hours_per_week * education_num.

* Rationale: High income is typically a function of both the quantity of work (hours) and the quality/level of skill (education). This interaction feature captures the combined effect of labor and intellectual capital.


##Numerical Transformation
##**Feature Selected:**
capital_gain
##**Transformation Applied:**
Log Transformation (log1p)
##**Rationale:**
 The capital_gain feature is highly right-skewed (Skewness $\approx 11.95$). This indicates that most people have zero capital gains, while a few have very high values. Such skewness can negatively impact model performance, especially for linear models. Applying log1p ($log(1+x)$) compresses the range and reduces the skewness (reducing it to $\approx 3.09$), making the distribution more manageable for the algorithm.